tabular/src/autogluon/tabular/models/widedeep_nn/widedeep_nn_model.py

```diff
@@ -0,0 +1,289 @@
+                time_elapsed = time.time() - start_time
+                time_left = time_limit - time_elapsed
+                if time_left <= time_limit * 0.7:  # if 30% of time was spent preprocessing, likely not enough time to train model
+                    raise TimeLimitExceeded
+            else:
+                time_left = None
+
+            early_stopping = EarlyStoppingCallbackWithTimeLimit(
+                monitor=monitor_metric,
+                mode='auto',
+                min_delta=params['early.stopping.min_delta'],
+                patience=params['early.stopping.patience'],
+                time_limit=time_left,
+                best_epoch_stop=best_epoch_stop
+            )
+
+            if num_cpus is not None:
+                system_params = {'num_workers': num_cpus}
+
+            if num_gpus is not None:
+                # TODO: Control CPU vs GPU usage during inference
+                if num_gpus == 0:
+                    system_params['device'] = 'cpu'
+                else:
+                    # TODO: respect CUDA_VISIBLE_DEVICES to select proper GPU
+                    system_params['device'] = 'cuda'
+
+            trainer = Trainer(
+                model,
+                objective=objective,
+                metrics=[m for m in [nn_metric] if m is not None],
+                optimizers=tab_opt,
+                lr_schedulers=tab_sch,
+                callbacks=[model_checkpoint, early_stopping],
+                verbose=kwargs.get('verbosity', 2),
+                **system_params
+            )
+
+            trainer.fit(
+                X_train=X_train,
+                X_val=X_valid,
+                n_epochs=params['epochs'],
+                batch_size=batch_size,
+                val_split=val_split,
+            )
+            best_epoch = model_checkpoint.best_epoch
+            trainer.model.load_state_dict(torch.load(model_checkpoint.old_files[-1]))
+
+        self.model = trainer
+
+        # TODO: add dynamic epochs selection
+        self.params_trained['epochs'] = params['epochs']
+        self.params_trained['best_epoch'] = best_epoch
+
+    def _predict_proba(self, X, **kwargs):
+        X = self.preprocess(X, **kwargs)
+        X = self.missing_filler.transform(X)
+        X_tab = self._tab_preprocessor.transform(X)
+        if self.problem_type != REGRESSION:
+            preds = self.model.predict_proba(X_tab=X_tab)
+        else:
+            preds = self.model.predict(X_tab=X_tab)
+            preds = self.y_scaler.inverse_transform(preds)
+
+        if self.problem_type == BINARY:
+            return preds[:, 1]
+        else:
+            return preds
+
+    def __get_batch_size(self, params):
+        batch_size = params['bs']
+        # SAINT need larger batches because it is using information between rows
+        if params['type'] == 'SAINT':
+            batch_size *= 2
+        return batch_size
+
+    def __prepare_datasets_before_fit(self, X, y, X_val, y_val):
```

Comment:
a cleaner name could be `_preprocess_fit`

Reply:
renamed